### ISSUE TITLE
TypeError: Conn.Query is not a function

### ISSUE THREAD

**Issue by sunalive** [davidjconnelly/trongate-framework#107]
I just installed Trongate desktop and I'm trying to create a new app.  When I go past the MySQL credentials, I get the error "TypeError: Conn.Query is not a function" and the app resets.  I'm using Debian Linux (Peppermint OS) and have LAMP installed in the system.  How do I solve this?

**Comment by DaFa66** [comment]
Hi sunalive,

This error is usually one of two things and both are to do with your database settings.

firstly, you should have you MySQL or MariaDB **running** on port 3306 or what ever you have set in the database connection details
```
{
    "host" : "127.0.0.1",
    "port" : 3306,
    "user" : "root",
    "password" : "",
    "database" : "your_app"
}
```
Ensure the user name and password are correct.  Also the database name does not already exist.

secondly, it could be a permissions thing with your database users.  You can do all this in phpMyAdmin or open a terminal and cd into your MySQL bin folder and execute the following:
```
 mysql.exe -uroot -p
 ALTER USER 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY '';
 FLUSH PRIVILEGES;
 exit
```

**Comment by sunalive** [comment]
Thanks for the response. I am able to login to phpmyadmin using the same username and password.  I'm able to connect normally to mysql from my other apps.  The DB has not been already created for the Trongate project. I'm using the standard LAMP stack (not XAMPP) and I have mysqld listening to 3306 and 33060 with document root under the user with 777 permissions. Still no luck :(

**Comment by DaFa66** [comment]
Are you logging into phpMyAdmin with user root and a password?
Did you execute the queries I mentioned above on user root?
```
 ALTER USER 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY '';
 FLUSH PRIVILEGES;
```
Not quite understanding "I have mysqld listening to 3306 and 33060".  It will be one or the other, have you tried both in the desktop app?

You can also create a new user like so and try using this instead of root:
```
 CREATE USER ***@***.*** IDENTIFIED BY 'qwertyuiop';
 GRANT ALL PRIVILEGES ON trongate.* TO 'trongate_user'@'localhost';
 FLUSH PRIVILEGES; 
```
I'm still leaning to a permissions issue with your setup, let me know how you get on.

**Comment by sunalive** [comment]
Yes, I have configured my MySQL install with a local user other than root, say 'mydbuser' and a unique password with all privileges.  I ran the above `GRANT & FLUSH` commands but the result is still the same.  My `netstat -tunlp` shows below:
```
tcp6       0      0 :::33060              :::*                    LISTEN      3050693/mysqld      
tcp6       0      0 :::3306               :::*                    LISTEN      3050693/mysqld     
tcp6       0      0 :::80                 :::*                    LISTEN      1295204/apache2   
tcp6       0      0 :::443                :::*                    LISTEN      1295204/apache2   
```

I tried changing the port in the app as you suggested and now I get a different error. 

![trongate-error](https://user-images.githubusercontent.com/2279283/186303435-b1b0ec47-5252-41ef-9ac4-b9742b929035.png)

**Comment by DaFa66** [comment]
- hmm that's interesting - did that JavaScript error appear on using a port of 33060 in the desktop app?
- Have you ever tried using root as the username, with or without a password for you db connection?
- Have you tried just downloading [Trongate from GitHib](https://github.com/davidjconnelly/trongate-framework), configure both config.php and database.php then creating a db manually or grabbing a module off the [module market](https://trongate.io/module-market) that has a db (like "Simple Member") to see if your stack is working as expected?
- Have you tried launching the desktop app with elevated privileges?

Then we can narrow down the issue being something not right with your setup and the desktop app.

I'm on Windows (with Linux via WSL2 [Ubuntu 20.04]) and also a Mac, so I can't reproduce your problem unless I create a vm of Peppermint OS > could you please tell my what versions of OS and LAMP stack you are using (assuming you are using [bitnami](https://bitnami.com/stack/lamp/installer) ) and any other setup you can make me aware of?  I'll then try recreate your issue.

**Comment by davidjconnelly** [comment]
This has come up a few times and we've usually found a solution.  Sometimes it happens if there are two instances of the desktop app running at the same time.  

Below is the URL from the Help Bar where this issue was previously raised (and solved):

https://trongate.io/help_bar/thread/FyNgZYs2wFg2

Below is another solution from a very old forum post:

_Hi Davcon, thanks for the reply. First solution to problem was to use a user that is already in the mysql users list (not root without pw - does not work here) or create new. 
Next problem was choosing a location - in my case it is in var/www/html/ (typically ubuntu/linux).
Error is Desktop App could not install in that location. It seems Desktop App cannot write there.
I dont use xampp._

**So.... I'm going to take a wild guess and assume that this problem of yours is being caused because the desktop app cannot find MySQL on your computer.  Just a guess.**

THREE POSSIBLE SOLUTIONS (easiest to most difficult)

1).  Use XAMPP.  Problem solved.  

2).  Hire somebody who is a guru of the Linux operating system and who'll know how to tear into the CLI and get it working.  I know a guy with those kinds of skills but you'll have to pay him - probably about 50 to 100 dollars in cash or crypto.

3).  Arrange a call with me on Skype.  I can set up a laptop with the same operating system as you.  I'd be willing to install the same things as you and attempt to recreate the error.

**Comment by sunalive** [comment]
Thanking both for your time. 
 
To answer all questions:

1. Peppermint OS is a standard Debian distribution with XFCE as DE. Nothing fancy, it is pretty much the same as Xubuntu without Ubuntu bloat. I like keeping things to the bare metal which is why I'm most interested in Trongate. 
2. I use a standard LAMP install (no packaged solution like Bitnami) - Apache, Php (as Apache module), MySQL (no Maria DB) and PhpMyAdmin. Similar to this link: [https://tecadmin.net/install-lamp-ubuntu-20-04/](https://tecadmin.net/install-lamp-ubuntu-20-04/), just no firewall in my case.
3. Apache and MySQL are global services with path enabled. I can run mysql commands from the Terminal globally without sudo.
4. I have `secure_mysql_installation` enabled with a regular user (other than root) configured with password and granted all rights (*.*).  This is the user I use for all my development. MySQL is accessible via PhpMyAdmin from any browser with the username and password.
5. The document root is redirected to `/home/<user>/Projects/www` instead of the default `/var/Apache2/htdocs`.  This folder is given 777 access so everyone can create folders/files without sudo. The <user> is also a member of www-data group (which Apache uses for webserver) so no permission conflicts.
6. XAMPP, if I set up, will also function the same way as the standard LAMP install since XAMPP requires root/sudo to start and stop.  The doc root will be /opt/lamp/htdocs which I'll have to route to a local user directory (like `/home/<user>/Projects/www`) or give 777 access to the `/opt/lamp/htdocs` folder.

I checked the solution in the link David provided, they had used MariaDB and disabled secure_mysql_installation.  I'll try with a fresh XAMPP install and check later today.  Alternate is to fire up a VM with Windows and get this running from there but that defeats the purpose and will be slow for my development. 

At this point, I feel there could be issues with the NodeJS access to Apache/MySQL which Trongate seems to use at the backend.  I don't have NodeJS installed as I don't do any JS app development.  Also not sure which user Trongate uses for its access (assume it should be the same as the current frontend user).

@davidjconnelly  Not sure which timezone you are in, I work in US EST (New York) zone. Please let me know your convenience.  I'll try to set up something if my XAMPP trial does not go well.

**Comment by davidjconnelly** [comment]
Okay.  That's all cool.

Just to let you know, I too use Peppermint OS.  It's my favourite operating system (or whatever).  So, I'll certainly be glad to do a screenshare on Skype and see if we can sort this.

Generally, 7pm London time works for me.

**Comment by sunalive** [comment]
I tried two things:

1. Clone the Github repo and see if it worked.  It did, as expected, after I changed the config to include my document_root and my db credentials.
2. Next, I disabled my LAMP install and installed XAMPP (with default options). I changed the document_root to my Project folder under my user. Left the DB credentials to default (root and blank) and tried the desktop app. It worked!  This time it created the basic app and a blank database with the app name.

So, I guess the desktop app did not like MySQL (which should not be the case), perhaps the node-modules included in the desktop app do not support MySQL or may be something else.  You might want to take a look at this.

For now, I can live with this. I would need to switch servers every time for my development or I need to migrate all my existing DBs to XAMPP.  However, I would love this to work with the default LAMP install, which is how most Linux users do their development. 

Thanks for all your support.

**Comment by epicWhite** [comment]
https://github.com/davidjconnelly/trongate-framework/issues/46

ср, 24 авг. 2022 г. в 19:57, sunalive ***@***.***>:

**Comment by DaFa66** [comment]
Hi sunalive,

Thanks for all that info, When I get a moment I'll try create a vm of Peppermint with XFCE as DE to see if I can recreate the problem.

As DC mentioned before, we have had a few people with the same error before, all of which had been resolved.  It seems to be a permission or setup thing with MySQL.  Be assured Trongate works fine with MySQL, I personally prefer MariaDB - but each to their own.

Good to see you have The Trongate Desktop App working with Xampp too.  And don't be confused with the Trongate Framework which has nothing to do with Node.js and the Desktop app, which is written with Electron.

Talking about LAMP stacks, I've been playing around with [devilbox](http://devilbox.org/) a Docker LAMP development stack with valid https support.  Works on all platforms, just clone the repo, add and .env and off you go with your own intranet and the ease to swap in and out versions of apache, nginX (note: nginX needs to be configured to work with Trongate, as by default uses .htaccess modrewtites with Apache), php v5.2 to 8.2, MySQL v5.5 to 80, MariaDB 5.5 to 10.7, and other soon to be supported databases, PgSQL, Mongodb and others all at your figure tips.  I may do a YouTube video on it and using it with Trongate soon.

**Comment by davidjconnelly** [comment]
Since this one appears to have been solved by switching to XAMPP, I'm going to close it.  If there's anything further, please do feel free to re-open.  Thanks for all the input with this one - especially Dafa.

**Comment by sunalive** [comment]
Well, actually it did not. XAMPP just created an empty database, it did not create the basic tables (like users).  So I ended up ditching the XAMPP and Trongate app instead, trying with the Git clone since Trongate app seems to just copy the basic module skeleton when we create one.  The framework itself is working fine with MySQL as I had mentioned before, it is just the app that has a problem and I think it is just in the Linux platform. I'll try the app again once you were able to fix the issue on Linux, will work with the Git clone till then.  Thanks for your time and support.

**Comment by trongate** [comment]
Well, please do keep in mind that I too like using Linux Peppermint.  As a matter of fact, I'm on it now.  So, if you want to meet on Skype for a screenshare and an opportunity to get it all working then the offer is there.  Just let me know.  -  DC

**Comment by DaFa66** [comment]
@sunalive your understanding of how the Desktop app works when creating a new app is not quite correct.

1. Yes, when you created a new app with the desktop app it created an empty database.  This is as expected, and not an error or malfunction in the process.
2. It is when you create a module or import one from the Module-Market with the Desktop app that it will create the standard tables:
```
Tables_in_sunalive       
-------------------------                   
trongate_administrators  
trongate_comments        
trongate_tokens          
trongate_user_levels     
trongate_users      
```
I haven't had the time to create a VM with Peppermint as yet, but still intend to try find out why you get a db connection error.  I still strongly believe it is an issue with your setup and not a problem with the Desktop app as it seams to be theme with Linux and Mac users, and usually comes down to a permission problem.  Not to mention others can get the Desktop App working fine on Peppermint or other Linux flavours with MySQL using stacks other than Xampp.

I was wondering do you use Docker?  If so, you could try a great repo called http://devilbox.org/.  I have it installed on Windows11 via WSL2 (Ubuntu 20.4) for testing different versions of Apache/PHP/MySQL or MariaDB and it works flawlessly with the Trongate Desktop App and the Framework.  It's also a great tool to test and expand Trongate to work with NginX and other database types without the hassle of installing and configuring them.

Cheers,
DaFa

**Comment by sunalive** [comment]
@DaFa66 No, I don't use Docker. I'm a hobby developer with a different main profession. I'll try to create the modules and check later today. But as per the documentation at Trongate, I understood that the basic tables are created when you install and create an app. May not be so. 

@trongate I'll try DaFa's suggestion and get back if I need your help. 

Again, thanks to you both for the good work.

**Comment by sunalive** [comment]
At last, I could spend enough time on this one. DaFa's suggestion worked! I installed the Simple User module and could see those tables created. I initially had issues with Trongate app not installing the modules (it just asked me the location to download instead of downloading and giving me an option to install). I figured that was about the Linux permissions, it worked once I fixed the app folder permissions to 777. 

I'm still playing around and trying to understand this framework. At the first look, I feel some code redundancies (IMO) that we are having to define stuff in multiple places (for eg. the theme name, which is being defined in every module view).  So refactoring/reorganizing may become an issue when the app grows big. I tried to remove the theme definition in the module view thinking a default would take over automatically but it did not happen that way.  Guess I need to read the documentation in detail.  Also, I feel some basic stuff can be parametrized and built into the engine (like user module which is going to be used in every implementation).

Overall, I'm very excited that things started working (albeit only in XAMPP for now).  I'll try to implement an application with this and see how it goes. Thank you both for your support.
